Add operation-based rate limiting

afterburn opened the pull request:
**Summary**
This PR replaces path/method/query-based rate limiting with operation-based configuration that maps directly to the relay's three core operations: publish, resolve, and resolve_most_recent. Simplifies configuration while maintaining all existing functionality.

**Configuration Example**
```toml
[relay]
[[relay.rate_limits]]
operation = "resolve"
quota = "10r/s"
burst = 20

[[relay.rate_limits]]
operation = "resolve_most_recent"
quota = "2r/s"
burst = 5

[[relay.rate_limits]]
operation = "publish"
quota = "2r/s"
burst = 10
whitelist = ["127.0.0.1", "192.168.1.0/24"]  # optional, both ips and subnets supported
```

**Operations in this toml file automatically map to:**
- publish → PUT requests (publishing records)
- resolve → GET requests without most_recent=true (cached lookups)
- resolve_most_recent → GET requests with most_recent=true (DHT queries)

If no config.toml file is present, system will use defaults, so this is not a breaking change.

afterburn commented:
@SeverinAlexB resolved your feedback:
- Removed kb, mb, gb quota units, now always uses either r/s or r/m as requested
- Changed default rate limits to something saner
- Added `behind_proxy` flag in config toml, which allows users to respect/ignore proxy headers
- Cleaned up IP extraction, only happens once per request now
- Added operation 'Index' so users can specify rate limiting for root path

SeverinAlexB commented on relay/src/handlers.rs:
```diff
@@ -63,12 +66,19 @@ pub async fn put(
         .map_err(|error| Error::new(StatusCode::BAD_REQUEST, Some(error)))?;
 
-    if let Some(signed_packet) = state.client.resolve(&public_key).await {
+    let signed_packet = if query.most_recent {
+        state.client.resolve_most_recent(&public_key).await
+    } else {
+        state.client.resolve(&public_key).await
+    };
```
Do we need a way to fully disable resolve_most_recent? A public relay might not want to allow it at all?

What do you think? Do we need a `resolve_most_recent_enabled = true/false` config?

SeverinAlexB reviewed (commented):
Two comments, otherwise looks good.

In the Pubky Core team, we have a convention on how to format PR title. We stick to the [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/).

So for this PR, the title could be: feat(relay): Operation-based rate limiting + Resolve most recent


<img width="1932" height="1302" alt="CleanShot 2025-09-25 at 10 37 51@2x" src="https://github.com/user-attachments/assets/faf99515-9ed4-467f-8746-877f961ff95c" />

afterburn commented:
@SeverinAlexB 

- If no rate limits are defined in `config.toml`, rate limiting will be disabled completely
- Added `Index` to `Operation` enum so calling `GET http://0.0.0.0:6881/` doesn't trigger the incorrect rate limit
- Added `resolve_most_recent_enabled` flag so users have the ability to completely ignore the `most_recent=true` query parameter (defaults to `false` for backward compatibility)

afterburn commented:
@SeverinAlexB @SHAcollision Restored DHT rate limiting that was previously removed. Since the DHT can be accessed directly, keeping rate limiting in place is necessary. The implementation is now fully backward compatible. Existing `[rate_limiter]` configurations continue to work as before. New `[dht_rate_limit]` and `[http_rate_limit]` options are opt-in. By default, it falls back to the legacy IP-based limiter unless one of the new configs is defined. Also added a comprehensive test to cover the new behavior and ensure backward compat.

86667 commented on pkarr/Cargo.toml:
```diff
@@ -1,12 +1,13 @@
 [package]
 name = "pkarr"
-version = "5.0.3"
+version = "6.0.0-rc.0"
```
if so, maybe we can remove the config backwards compatibility logic

86667 commented on relay/src/rate_limiter/layer.rs:
```diff
@@ -0,0 +1,598 @@
+                tracing::debug!("Rate limit of {} exceeded for {:?}", limit.limit.quota, key);
+                return Box::pin(async move {
+                    Ok(
+                        HttpError::new(StatusCode::TOO_MANY_REQUESTS, Some("Rate limit exceeded"))
```
`Retry-After` header always nice for 429s

86667 reviewed (commented):
Few comments 👍 

Also we could have quite a few more unit tests. All behaviour may be tested somewhere but unit tests inline with the code they test are pretty much always worth it these days with how fast AI can write them IMO
